### PR TITLE
[Clustering] Support for clustering of a subclassed model.

### PR DESCRIPTION
**Pull request by wwwind**
This PR adds support for clustering of a subclassed model. Added an example for clustering of a subclassed model taken from the tutorial: https://www.tensorflow.org/tutorials/quickstart/advanced

### CODE REVIEW

File: tensorflow_model_optimization/python/core/clustering/keras/cluster_integration_test.py [hunk 1]
```diff
@@ -160,6 +184,105 @@ def clusters_check(stripped_model):
...
+    # The model should be trained a little bit.
+    clustered_model.fit(x=self.dataset_generator(), steps_per_epoch=1)
+    stripped_model = cluster.strip_clustering(clustered_model)
+
+    nr_unique_weights = len(np.unique(stripped_model.layers[0].\
+        trainable_weights[0].numpy().flatten()))
+    self.assertLessEqual(nr_unique_weights, self.params["number_of_clusters"])
+
+  @keras_parameterized.run_all_keras_modes(always_skip_v1=True)
+  def testEndToEndSubclassedModelTwoLayers(self):
```
Comment (wwwind) [inline tensorflow_model_optimization/python/core/clustering/keras/cluster_integration_test.py:214]: This test re-produces the approach tested here: https://github.com/tensorflow/model-optimization/pull/554

File: tensorflow_model_optimization/python/core/clustering/keras/cluster_integration_test.py [hunk 2]
```diff
@@ -160,6 +184,105 @@ def clusters_check(stripped_model):
...
+        trainable_weights[0].numpy().flatten()))
+    self.assertLessEqual(nr_unique_weights, self.params["number_of_clusters"])
+    nr_unique_weights = len(np.unique(subclassed_model.layers[1].\
+        trainable_weights[0].numpy().flatten()))
+    self.assertLessEqual(nr_unique_weights, self.params["number_of_clusters"])
+
+  @keras_parameterized.run_all_keras_modes(always_skip_v1=True)
+  def testEndToEndSubclassedModelAsDeepLayer(self):
+    """Test End to End clustering for the model with the layer as a subclass model."""
+    # This case is not supported currently.
```
Comment (wwwind) [inline tensorflow_model_optimization/python/core/clustering/keras/cluster_integration_test.py:274]: this case will be enabled later once the current approach is approved

File: tensorflow_model_optimization/python/core/clustering/keras/cluster_test.py [hunk 3]
```diff
@@ -385,6 +385,48 @@ def testClusterFunctionalModelPreservesBuiltState(self):
           json.loads(clustered_model.to_json()))
     self.assertEqual(loaded_model.built, True)
 
+  @keras_parameterized.run_all_keras_modes
```
Comment (wwwind) [inline tensorflow_model_optimization/python/core/clustering/keras/cluster_test.py:388]: it looks that the block of these tests  has not been upstreamed. i can remove it from my PR as they are not relevant

**Comment by wwwind** [comment]
Hi @alanchiao Please take a look at this PR: it enables clustering for a subclassed model.
when the whole model is passed, it is wrapped, so we have control over 'build/call' functions of layer:
this is the [wrapper](https://github.com/tensorflow/model-optimization/pull/571/files#diff-3dc12136199cd67cae87bb97c608a601a750750a3219b95f4fefc5efa4510e17R349)
the approach has been tested on the subclassed model from this [tutorial](https://www.tensorflow.org/tutorials/quickstart/advanced)
Any comments/suggestions are much appreciated. Thanks!

**Comment by alanchiao** [comment]
Haven't looked into this PR in detail yet, but the main thing is that we need to see how well this approach actually works, from the doc I had shared. The comments on the side of the doc need to be resolved in a satisfactory manner.

Ultimately, when any person asks for subclassed model support, they're really looking for a task or model to be supported (Object Detection, NLP) and until we are confident it works reasonably with at least one of those using a standard implementation (e.g. TF Official Models), then it doesn't add significant value.

**Comment by wwwind** [comment]
Hi @alanchiao Yes, all issues that are mentioned in the doc are addressed. The clustering case is easier, because we cluster the trained model always. The example on mnist demonstrates that we use the same Clustering API - I just removed summary() and save/load() into HDF5 as they are not supported by the subclassed model.
On my list TODO is to cluster MobileBert.
Thanks.

**Comment by alanchiao** [comment]
@wwwind : it's not immediately obvious why the fact that clustering is always applied to a pre-trained model makes a difference. The suggested problem comes from setting the attributes of the layers of a subclassed model to the optimized versions of those layers, which seems independent of the model optimization technique.

That said, I haven't fully understood the risks / issues yet (e.g. I questioned further on the second issue for instance).

If you and others feel comfortable with moving forward with it though, you can choose to do so. It may be good to say that the approach hasn't been tested on more complex cases such as Object Detection in Official Models.

**Comment by wwwind** [comment]
it is not relevant anymore

**Pull request closed without merging**